Document environment variables

nrwahl2 opened the pull request:
Still need to finish up by documenting client variables either in Pacemaker Explained or in Pacemaker Administration.

Ref T637

nrwahl2 commented:
Formatting is strange in the new section of Pacemaker Explained. I'm using the same widths as the cluster options table, but in the new node options table the Description column is much wider and causes the other columns to be narrow and wrap text.

nrwahl2 commented on etc/sysconfig/pacemaker.in:
```diff
@@ -275,6 +275,16 @@
 #
 # Default: PCMK_ipc_buffer="131072"
 
+# PCMK_cib_timeout (Advanced Use Only)
```
Alternatively this could go to its own section. It's technically related to inter-process communication but isn't part of the IPC configuration.

nrwahl2 commented on doc/sphinx/Pacemaker_Explained/node-options.rst:
```diff
@@ -0,0 +1,485 @@
+     - 1024
+     - *Advanced Use Only:* Set a lower bound on the bit length of the prime
+       number generated for Diffie-Hellman parameters needed by TLS connections.
+       The default is no minimum.
```
unclear how this aligns with the default as "1024"

nrwahl2 commented on doc/sphinx/Pacemaker_Explained/node-options.rst:
```diff
@@ -0,0 +1,485 @@
+     - 2048
+     - *Advanced Use Only:* Set an upper bound on the bit length of the prime
+       number generated for Diffie-Hellman parameters needed by TLS connections.
+       The default is no maximum.
```
ditto

kgaillot replied:
ditto :)

nrwahl2 commented on doc/sphinx/Pacemaker_Explained/node-options.rst:
```diff
@@ -0,0 +1,485 @@
+
+       PCMK_authkey_location
+     - :ref:`text <text>`
+     - /etc/pacemaker/authkey
```
Default paths like this one will actually depend on what the `configure` script does to `pacemaker.in`. Not sure how else to represent it besides leaving the default unspecified.

kgaillot replied:
`doc/sphinx/conf.py.in` can define variables (`rst_prolog`)

nrwahl2 replied:
That's cool (I didn't know) but only one version can be hosted on the ClusterLabs website.

kgaillot replied:
Ah right, I wasn't thinking. Still worth doing for distro doc packages. We could maybe add something parenthetical like "may be different in your build". It sure would be handy for someone to be able to list all options with defaults in their build. ;)

nrwahl2 replied:
Makes sense in theory though I'm not sure off hand how we'll do it differently for the distro doc packages. Maybe something in autoconf. Or we can use the variables in `rst_prolog` all the time to make things simple. That's probably the way to go. In practice we'll probably be building the ClusterLabs docs on Fedora or similar for a while, so the default directories are unlikely to change between builds.

kgaillot replied:
`conf.py.in` is already configure-substituted, so `rst_prolog` just needs to use the configure variable and everything will automatically flow. By distro packages, I mean each distro can configure Pacemaker how they want, and if they ship a doc package, it will automatically show their configured value.

nrwahl2 replied:
It's not playing nicely for me yet. With the following substitution defined:
```
.. |CONFIGDIR| replace:: @CONFIGDIR@
```

It's giving me a literal `@CONFIGDIR` in the doc. I've tried a few different syntax arrangements. Will revisit in a bit.

kgaillot replied:
Doh, I forgot conf.py.in is not a configure .in file but handled specially in its Makefile -- should be able to add something there

nrwahl2 replied:
Thanks! I was able to get the substitution working that way. `Makefile.am`:
```
$(BOOKS:%=%/conf.py): conf.py.in
        $(AM_V_at)-$(MKDIR_P) "$(@:%/conf.py=%)"
        $(AM_V_GEN)sed                                                  \
                ...
                -e 's#%CONFIGDIR%#@CONFIGDIR@#g'                        \
```

`conf.py.in`:
```
rst_prolog="""
...
.. |CONFIGDIR| replace:: %CONFIGDIR%
"""
```

There is a minor annoyance. Substitution references in RST are considered inline markup and cannot be nested inside an inline literal (double backticks). So where I'm trying to say ` ``|CONFIGDIR|/pacemaker`` ` in the body of a paragraph, it will not be possible as such. There is no workaround (other than making the books themselves configure-replaceable, which isn't worth it) -- see for example https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#inline-markup

kgaillot replied:
Yeah that's not too big a deal. If it really bothers you, you could add a second replacement for the formatted subdirectory name.

nrwahl2 commented on doc/sphinx/Pacemaker_Explained/node-options.rst:
```diff
@@ -0,0 +1,547 @@
+
+       PCMK_cib_timeout
+     - :ref:`nonnegative integer <nonnegative_integer>`
+     - the maximum of 30 and (10 * ((number of nodes in the cluster) + 1))
```
true but sloppy

kgaillot commented on etc/sysconfig/pacemaker.in:
```diff
@@ -212,6 +212,15 @@
 
+# PCMK_remote_pid1 (Advanced Use Only)
+#
+# Specify behavior when the Pacemaker Remote process is not running as PID 1.
```
I'm thinking:

When a bundle resource's "run-command" option is left to default, Pacemaker Remote runs as PID 1 in the bundle's containers. When it does so, it loads environment variables from the container's /etc/pacemaker/pcmk-init.env, and performs the PID 1 responsibility of reaping dead subprocesses. This variable controls whether those behaviors are performed when Pacemaker Remote is not running as PID 1. It is intended primarily for developer testing but can be useful when "run-command" is set to a separate, custom PID 1 process that launches Pacemaker Remote.

kgaillot commented on etc/sysconfig/pacemaker.in:
```diff
@@ -212,6 +212,15 @@
+# Specify behavior when the Pacemaker Remote process is not running as PID 1.
+# * If set to "full", Pacemaker Remote behaves as though it's running as PID 1.
+# * If set to "vars", Pacemaker Remote loads environment variables from
+#   /etc/pacemaker/pcmk-init.env, while all other PID 1 behavior is disabled.
```
* If unset, Pacemaker Remote performs neither behavior when it is not PID 1.
* If set to "full", Pacemaker Remote performs both PID 1 behaviors even when it is not PID 1.
* If set to any other value (such as "vars"), Pacemaker Remote loads environment variables from /etc/pacemaker/pcmk-init.env even when it is not PID 1, but does not read dead subprocesses.

kgaillot commented on etc/sysconfig/pacemaker.in:
```diff
@@ -305,6 +305,16 @@
+# PCMK_cluster_type (Advanced Use Only)
+#
+# Specify the type of cluster layer. Currently only corosync is supported, and
+# pacemakerd overrides this environment variable by setting it to "corosync".
```
Specify the cluster layer to be used. If unset, Pacemaker will detect and use a supported cluster layer, if available. Currently, "corosync" is the only supported cluster layer. If multiple layers are supported in the future, this will allow overriding Pacemaker's automatic detection to select a specific one.

kgaillot reviewed (commented):
Soon users won't be flying blind :)

nrwahl2 commented on etc/sysconfig/pacemaker.in:
```diff
@@ -212,6 +212,15 @@
+# * If set to "vars", Pacemaker Remote loads environment variables from
+#   /etc/pacemaker/pcmk-init.env, while all other PID 1 behavior is disabled.
+#
+# Default: PCMK_remote_pid1="vars"
```
This is true, but there's also no special handling of `"vars"` whatsoever. That was by design. `"vars"` behaves exactly like unset -- see PR description of #1891. So it's effectively the default.

My idea here was to treat this as an enumeration and nudge users toward setting either `"vars"` or `"full"` if they set this at all -- pretending that other values are invalid.

kgaillot replied:
Unset and anything-but-full are handled differently. If the environment variable is unset, we return immediately without doing anything. If it's anything besides "full", we do just the vars.

kgaillot replied:
Since this is a relatively recent addition and an extremely narrow use case, I wouldn't mind making it a true enumeration and coming up with three distinct values, defaulting to one of them.

nrwahl2 replied:
Ah yep, I misread, thanks for catching it.

True enumeration sounds good to me.

nrwahl2 commented on doc/sphinx/Pacemaker_Explained/node-options.rst:
```diff
@@ -0,0 +1,547 @@
+Per-node Configuration
+--------------------------
```
I was on the fence, so that's fine with me.

I went with a separate section because the Nodes section is about Pacemaker's conception of nodes -- defining them, setting node attributes in the CIB, etc.

These configuration options have nothing to do with "nodes". They're options that are applied locally on the machine (node) on which they're set. In my mind the emphasis is more that these are local configs.

kgaillot replied:
Hmm, true. It should be the first chapter after the introduction, maybe "Host-local options" or "Start-up options".

nrwahl2 replied:
They're applied only at startup, but I don't want to give the impression that they affect only startup behavior, so I'll go with host-local.

nrwahl2 commented on etc/sysconfig/pacemaker.in:
```diff
@@ -275,6 +275,16 @@
+# (10 * ((number of nodes in the cluster) + 1)).
+#
+# Default: PCMK_cib_timeout=""
+
```
> This is used to cap the computed value, rather than the computed value being the default of this.

It's whichever one is larger. It doesn't look like this caps the computed value.
```
calculated_timeout = QB_MAX(calculated_timeout, env_timeout);
```

My description is also wrong though. The actual timeout is effectively `max(30, PCMK_cib_timeout, calculated_timeout)`.

---

Yeah, I think a cluster option would be fine. If so, can we drop the env var, or keep for compatibility? It was never documented.

Not sure about `cluster-delay`. I agree that it's documented as being for inter-node communication, and it's used to set a `network_delay` member. However, it's not clear to me that it applies to "delays from any source." It is used as the controller's default timeout for cluster actions and resource actions (at least the latter may be local), if the configured timeout is <= 0.

This use case is for IPC with the local CIB manager via `controld_globals.cib_conn`, rather than inter-node. My feelings aren't strong though.

kgaillot replied:
Bah I read this one twenty times and still interpreted it wrong. Let's leave it undocumented and deprecate it via comment. I don't think we need it at all (even as a cluster option). With 2 or fewer active nodes, we'll use 30s. With more than 2, the formula already seems excessive to me -- there should never be a reason to raise it even higher.

nrwahl2 commented on etc/sysconfig/pacemaker.in:
```diff
@@ -81,6 +81,12 @@
 
+# PCMK_stderr (Advanced Use Only)
+#
+# Whether to send daemon log messages to stderr.
```
Should it be documented in sysconfig or not? I don't think sysconfig will be read if a user starts Pacemaker manually on the command line.

nrwahl2 replied:
I guess it could go in there as preparation for creating a "true" config file (T574). In the future, we'll likely parse the configuration file in a more sophisticated way than passing it to the shell to pull environment variables.

kgaillot replied:
Yeah, it's worth it just to have the documentation. Probably worth adding to the comment that setting it in this file is pointless.

nrwahl2 commented:
Updated to address review. The one thing I still can't figure out is how to set column widths on the table. No matter what values I set for `widths`, the column widths don't change, and I get a bunch of wrapping that I don't want.

I'm using the same values (`2 2 2 5`) as the Cluster-Wide Configuration section uses. I've also tried `auto` and `grid`, and I've tried removing the `longtable` class to see if that would even make a difference. It did not.

kgaillot commented on etc/sysconfig/pacemaker.in:
```diff
@@ -192,8 +192,8 @@
 # Remote connections. This file must be readable by Pacemaker daemons (that is,
-# it must allow read permissions to either the hacluster user or the haclient
-# group), and its contents must be identical on all nodes.
+# it must allow read permissions to either the @CRM_DAEMON_USER user or the
```
@

nrwahl2 replied:
whoops

kgaillot commented on doc/sphinx/Pacemaker_Explained/local-options.rst:
We can skip documenting G_SLICE through MALLOC_CHECK_ here since they're not actually Pacemaker options. They're useful in sysconfig just for developer purposes.

nrwahl2 replied:
True

kgaillot commented:
Looks great. Since this doesn't touch code significantly, let's do it against 2.1 once rc1 is out.

nrwahl2 commented:
@kgaillot Any ideas about the column widths? They're not going to be perfect but they are too narrow now IMO (except for the description column, which is too wide).

Despite using the same settings as the CIB properties table, these columns have different widths compared to that table, and I can't get them to change.

kgaillot commented:
> @kgaillot Any ideas about the column widths? They're not going to be perfect but they are too narrow now IMO (except for the description column, which is too wide).
> 
> Despite using the same settings as the CIB properties table, these columns have different widths compared to that table, and I can't get them to change.

It's your example in PCMK_trace_functions. The formatting is finicky. It respects the column sizes as much as it can, but it can only break words in certain ways, so if there's a long string that can't be broken, it winds up forcing a column width.

That particular example is the longest unbreakable string, and deleting it is enough to make it look decent on my browser. I'm not sure if something else might cause problems in another browser.

I would either "rename" the two functions shorter or drop one of them.

nrwahl2 commented:
Addressed review in first push, and then rebased on current main in the second push, hoping to resolve new coverity errors in code that I haven't touched :/

nrwahl2 commented:
retest this please

nrwahl2 commented:
the past two times it's just been opensuse build failures

kgaillot commented:
Looks good, needs another rebase

nrwahl2 commented:
Rebased on current 2.1
